docs/enhancements/add-external-providers.md

```diff
@@ -0,0 +1,182 @@
+    //...
+}
+```
+
+##### Step 3: Implement registry
+
+Implement the registry pattern so providers can register with their names.
+When we will detect the a certain provider then we will register it.
+If no provider is detected then we register the Bare-Metal provider.
+
+##### Step 4: Implement providerDetector method
+
+Implement a function which determines which provider should be used.
+Each provider should implement its own detection logic.
+For example:
+if all control plain nodes are running on oVirt
+(i.e have [SystemVendor.Manufacturer](https://github.com/openshift/assisted-service/blob/master/models/system_vendor.go)
+== oVirt OR RHV) then register the oVirt provider.
+
+##### Step 5: Add calls to the provider functions
+
+When ever we need to call some provider specific function we should use the provider interface instead of a switch case.
+
+##### Step 6: Implement Default Provider
+
+Will contain default implementation of most functions required by Provider interface. 
+It should be embedded in all providers to make adding new APIs easier.
+
+##### Step 7: Implement the oVirt provider and migrate Baremetal and VSphere providers
+
+Implement the oVirt provider and migrate existing providers to the Provider interface.
+
+#### Changes to the assisted-agent project:
+
+After the agent is running on the host, the provider should be determined from the OS details of the node.
+The [GetVendor](https://github.com/openshift/assisted-installer-agent/blob/master/src/inventory/system_vendor.go#L29-L45)
+method calculates if a node is virtual or not and populates the SystemVendor model which is later being used by the
+assisted service to determine if a node is running on a certain provider.
+We need to make sure that the provider is listed in the 
+[isVirtual](https://github.com/openshift/assisted-installer-agent/blob/master/src/inventory/system_vendor.go#L15-L19]
+list and that it can be detected using the fields in 
+[SystemVendor](https://github.com/openshift/assisted-service/blob/master/models/system_vendor.go).
+
+#### Changes to the Openshift installer project:
+
+A small change is required in the installer project.
+Since the openshift installer is generating a random InfraID the names of the Nodes(in the manifest) are different
+from the nodes the user provisions, and since we have a provider set then the provider will try to start new machines 
+due to the naming mismatch.
+Since the InfraID is random we can't tell the user to create the Nodes with specific names.
+Also, the InfraID is later used by the platforms in various ways to identify the cluster,
+for example in oVirt we use it to set an oVirt tag which helps us group the cluster resources in oVirt,
+and we rely on it in various places such as cluster destroy or CSI/Machine provider logic to filter cluster resources
+quickly and prevent unnecessary API calls.
+
+We need to add an env var to allow us to override the generated InfraID.
```

Comment:
The generation of the infraID is random but it can be known as in AWS UPI: https://github.com/openshift/installer/blob/master/docs/user/aws/install_upi.md#extract-infrastructure-name-from-ignition-metadata

Extracting the infra id would be preferred if it satisfies the needs. If it does not, let's discuss why.

Reply:
In the assisted installer the machines are provisioned before the ignition-configs or manifests are generated.
If we will go with this approach we will need to make manual modifications to the User Provisioned Nodes

Reply:
How does it work with UPI?

Reply:
Or we can chagne to the ignition config rendered by the installer :wink:

Reply:
@eranco74 it also affects the manifests a lot.
@avishayt On current RHV UPI we don't enable provider integration, we have a separate effort to document this as well, but essentially since the manifests and ignition can be created before there are any Nodes there is no problem with extracting the infraID and then creating the VMs with that infraID in the name